### PR TITLE
Add tier 6 unit tests for the dnsmasq package

### LINKED ISSUE #551: Unit tests for tier 6 (dnsmasq): dhcp_config, base, interface_handler, manager_mode, metalbox_mode, manager
## Background

Sub-issue of #524 (meta). Tier 6 covers the **dnsmasq** package under `files/netbox/dnsmasq/` — `dhcp_config.py`, `base.py`, `interface_handler.py`, `manager_mode.py`, `metalbox_mode.py` and `manager.py`. These modules turn the OOB-management view of NetBox (devices' OOB interfaces, OOB prefixes/VLANs, device types and tags) into dnsmasq DHCP host / MAC / range / option entries and write them into the generated inventory tree. No live NetBox is required.

This tier introduces **two new testing dimensions** that earlier tiers did not need:

1. **Filesystem writes.** `base.write_dnsmasq_to_device` and `dhcp_config.write_dhcp_ranges` create directories and write YAML files under `config.inventory_path`. The tests use pytest's `tmp_path` as the inventory root and assert on the files / directories actually produced (existing-dir vs existing-file-append vs new-dir branches).
2. **A fake `NetBoxClient` façade.** Unlike tiers 3-5 (which drove a bare pynetbox `api` stub), the dnsmasq code calls higher-level `netbox_client` methods — `get_oob_networks()`, `get_all_oob_prefixes()`, `get_device_oob_interface(device)`, `update_device_custom_field(...)` — *and* `netbox_client.api.*`. So the collaborator is a `MagicMock`-backed client whose method return values and `call_args` are what the tests assert on (`get_device_oob_interface` is covered by tier 8, here it is stubbed).

The six modules are bundled into one issue (analogous to how tier 4 bundled three modules into #546 and tier 5 bundled two extractors into #550) because they share the same `DnsmasqBase` / `DHCPConfigGenerator` collaborators, the same `tmp_path` + fake-client fixtures, and a single set of conftest factory additions. The issue is split into **seven reviewable parts** — matching the seven module checkboxes already tracked under Tier 6 in the meta (#524) — so the PR can be staged. `metalbox_mode.py` (~830 LOC) is the largest module and is split into Part 1 (pure / single-interface helpers) and Part 2 (the higher-level collectors and the `process_devices` orchestrator); if `process_devices` proves too large for one PR, Part 2 may be split further into bridged vs routed scenarios.

## Scope

Add `tests/unit/netbox/test_dnsmasq_dhcp_config.py`, `test_dnsmasq_base.py`, `test_dnsmasq_interface_handler.py`, `test_dnsmasq_manager_mode.py`, `test_dnsmasq_metalbox_mode.py` and `test_dnsmasq_manager.py` covering the modules listed below. Extend `tests/unit/netbox/conftest.py` with a few new shapes — `make_prefix(prefix, vlan=None)` for OOB network / prefix objects, a `make_vlan(vid, group=None)` extension and `make_vlan_group(name)` for the routed-group check, a `make_device(device_type=...)` extension for the MAC-entry device-type fallback, and a `make_fake_api(prefixes=...)` extension to back `ipam.prefixes.filter(tag=...)`. Add a small `make_dnsmasq_config(tmp_path, **overrides)` helper producing the `SimpleNamespace` config these modules read (`reconciler_mode`, `inventory_path`, `dnsmasq_lease_time`, `dnsmasq_switch_roles`). Keep every factory addition backward compatible (keyword-only, defaulted) so the tier 1-5 tests keep passing unchanged.

## Test targets

---

#### `DHCPConfigGenerator.__init__(config)` — `files/netbox/dnsmasq/dhcp_config.py:19`

- stores `config`

#### `DHCPConfigGenerator.generate_dhcp_host_entry(device, ip_address, mac_address, vlan_id=None, set_tag=None)` — `files/netbox/dnsmasq/dhcp_config.py:22`

- plain (no `set_tag`, no metalbox `vlan_id`) → `"{mac.lower()},{hostname},{ip}"`; the MAC is lowercased and the hostname comes from `get_inventory_hostname` (tier 1)
- explicit `set_tag` → `"...,{ip},set:{set_tag}"`; `set_tag` wins **even when `vlan_id` is also passed** (the `if set_tag` branch at line 48 precedes the vlan branch)
- `vlan_id` set **and** `config.reconciler_mode == "metalbox"` → `"...,{ip},set:vlan{vlan_id}"`
- `vlan_id` set but `reconciler_mode != "metalbox"` → plain entry (vlan ignored outside metalbox)
- hostname honours the `inventory_hostname` custom-field override vs `device.name` (assert both)

#### `DHCPConfigGenerator.generate_dhcp_mac_entry(device, mac_address)` — `files/netbox/dnsmasq/dhcp_config.py:55`

- **Priority 1** — `dnsmasq_dhcp_tag` custom field set → `"set:{custom_tag},{mac}"`
- **Priority 2** — no custom tag, device carries the `managed-by-ironic` tag → `"set:ironic,{mac}"`
- **Priority 3** — no custom tag, no ironic tag, `device.device_type.slug` present → `"set:{slug},{mac}"`
- none of the three available (no custom tag, no ironic tag, `device_type` falsy or slug falsy) → `None`
- MAC lowercased in every branch

#### `DHCPConfigGenerator.write_dhcp_ranges(netbox_client, prefix_tags=None)` — `files/netbox/dnsmasq/dhcp_config.py:85`

> Writes into `tmp_path / "group_vars" / "manager" / "999-netbox-dnsmasq-dhcp-range.yml"`.

- `get_oob_networks()` empty → early return, debug-logged, **no** file written
- IPv4 network, no `prefix_tags` → range `"{network_address},static,{netmask},{lease}"`; the file is written, `group_vars/manager` created (`mkdir parents/exist_ok`), top-level YAML key `dnsmasq_dhcp_ranges` is a list
- IPv6 network → **skipped** (`net.version == 6`, line 109)
- `prefix_tags` provided (routed mode): only prefixes present in the mapping are emitted (others skipped at line 119) as `"set:{tag},{addr},static,{mask},{lease}"`; a network whose prefix is absent from `prefix_tags` is omitted
- a malformed `network.prefix` that raises in `ipaddress.ip_network` → warning logged (line 130), loop continues, the remaining networks still processed
- all networks skipped / none valid → no file written (the `if dhcp_ranges` guard)
- lease time is read from `config.dnsmasq_lease_time`

---

#### `DnsmasqBase.write_dnsmasq_to_device(device, dnsmasq_data)` — `files/netbox/dnsmasq/base.py:20`

> Resolves the device dir via `host_vars/{hostname}*` glob under `tmp_path`; `hostname` comes from `get_inventory_hostname`.

- existing **directory** match (`host_vars/{hostname}/`) → writes `999-netbox-dnsmasq.yml` inside it (mode `w+`), YAML round-trips `dnsmasq_data`
- existing **single-file** match (`host_vars/{hostname}` is a file) → **appends** with a leading `"\n# NetBox dnsmasq\n"` separator (mode `a`), preserving prior content
- **no** match → creates `host_vars/{hostname}/` and writes `999-netbox-dnsmasq.yml`
- **multiple** glob matches (>1) → warning logged (line 35), nothing written, returns early
- hostname honours the `inventory_hostname` custom field vs `device.name` (glob pattern is `{hostname}*`)

---

#### `InterfaceHandler.get_virtual_interfaces_for_dnsmasq(device, netbox_client)` — `files/netbox/dnsmasq/interface_handler.py:15` (static)

- `api.dcim.interfaces.filter` raising → warning logged (line 39), returns `[]`
- no interfaces → `[]`
- non-`virtual` interface → skipped (the `type.value == "virtual"` gate)
- virtual without `tags` → skipped; virtual with tags but no `managed-by-osism` → skipped
- virtual + managed tag but **no** `untagged_vlan` → skipped
- virtual + managed + `untagged_vlan` → name appended (label preferred over name), debug-logged with the VLAN vid
- multiple qualifying interfaces → all appended, order preserved
- called as a static method (no instance needed)

---

#### `ManagerModeHandler.__init__(config)` — `files/netbox/dnsmasq/manager_mode.py:18`

- super-inits `DnsmasqBase` (stores `config`) and constructs a `DHCPConfigGenerator`

#### `ManagerModeHandler.process_devices(netbox_client, devices)` — `files/netbox/dnsmasq/manager_mode.py:22`

> `netbox_client.get_device_oob_interface(device)` returns `(ip, mac, vlan_id)`; stub it per scenario.

- MAC **and** IP → a host entry **and** a MAC entry generated; `update_device_custom_field(device, "dnsmasq_parameters", {...})` called once; `write_dnsmasq_to_device` receives `dnsmasq_dhcp_hosts__{hostname}` and `dnsmasq_dhcp_macs__{hostname}` keys
- MAC but **no IP** → no host entry (info-logged, line 63); MAC entry still generated; cached `write_params["dnsmasq_dhcp_hosts"]` stays `[]`
- **no MAC** → device skipped entirely (no write, no custom-field update)
- MAC present but `generate_dhcp_mac_entry` returns `None` (no tag source) → no `dnsmasq_dhcp_macs__*` key; the custom field is still written, but `write_dnsmasq_to_device` is **not** called when `dnsmasq_data` ends up empty (the `if dnsmasq_data` guard at line 90)
- `update_device_custom_field` returning a falsy value → warning logged (line 85)

#### `ManagerModeHandler.process_devices_readonly(devices)` — `files/netbox/dnsmasq/manager_mode.py:93`

- device with **no** cached `dnsmasq_parameters` → skipped (`continue`)
- `device.custom_fields is None` → `(... or {})` coercion, treated as no cache → skipped
- cached `dnsmasq_dhcp_hosts` → `dnsmasq_dhcp_hosts__{hostname}` written; cached `dnsmasq_dhcp_macs` → `dnsmasq_dhcp_macs__{hostname}` written; both present → both keys
- cached params present but both lists empty/absent → `dnsmasq_data` empty → no file written
- **no** API access: neither `get_device_oob_interface` nor `update_device_custom_field` is called (read-only path)

---

#### `DnsmasqManager.__init__(config)` — `files/netbox/dnsmasq/manager.py:17`

- constructs a `DHCPConfigGenerator`, a `ManagerModeHandler` and a `MetalboxModeHandler`

#### `DnsmasqManager.write_dnsmasq_config(netbox_client, devices, all_devices=None)` — `files/netbox/dnsmasq/manager.py:23`

> Patch `manager_handler` / `metalbox_handler` on the instance with `MagicMock`s and assert which path runs.

- `reconciler_mode == "manager-readonly"` → only `manager_handler.process_devices_readonly(devices)` called
- `reconciler_mode == "metalbox"` **and** `all_devices` truthy → `metalbox_handler.process_devices(netbox_client, devices, all_devices)` called
- `reconciler_mode == "metalbox"` but `all_devices` is `None`/empty → falls through to the manager branch (the `and all_devices` guard at line 38)
- any other mode (`"manager"`) → `manager_handler.process_devices(netbox_client, devices)` called

#### `DnsmasqManager.write_dnsmasq_dhcp_ranges(netbox_client, prefix_tags=None)` — `files/netbox/dnsmasq/manager.py:44`

- delegates verbatim to `dhcp_generator.write_dhcp_ranges(netbox_client, prefix_tags=prefix_tags)` (assert pass-through of both args)

---

#### `MetalboxModeHandler.__init__(config)` — `files/netbox/dnsmasq/metalbox_mode.py:19`

- super-inits `DnsmasqBase`, constructs a `DHCPConfigGenerator` and an `InterfaceHandler`

#### `_build_prefix_tag_mapping(oob_networks)` — `files/netbox/dnsmasq/metalbox_mode.py:24`

- only IPv4 prefixes are mapped; IPv6 prefixes are dropped (line 42)
- a prefix with a **unique** VLAN id → tag `"vlan{vid}"` (no suffix); a unique prefix with no VLAN → tag `"oob"`
- **duplicate** VLAN ids → suffixes appended in network-address-sorted order (`vlan100a`, `vlan100b`, …); duplicate no-VLAN prefixes → `ooba`, `oobb`, …
- ordering is deterministic — sorted by `network_address`, so the suffix assignment is independent of the input order (assert by shuffling the input)
- returns `{prefix_str: {"tag": ..., "network": <ip_network>, "vlan_id": ...}}`

#### `_index_to_suffix(idx)` — `files/netbox/dnsmasq/metalbox_mode.py:80` (static)

- `0 → "a"`, `25 → "z"`
- `26 → "aa"`, `27 → "ab"`, `51 → "az"`, `52 → "ba"`, `701 → "zz"` (the documented 0..701 range)

#### `_get_set_tag_for_ip(ip_address, prefix_mapping)` — `files/netbox/dnsmasq/metalbox_mode.py:93`

- an IP inside a mapped prefix's network → that prefix's tag
- an IP outside every mapped prefix → `None`
- an `ip_address` carrying a `/prefix` suffix → the mask is stripped before matching (line 103)

#### `_get_metalbox_oob_virtual_interface(device, netbox_client)` — `files/netbox/dnsmasq/metalbox_mode.py:109`

- **Priority 1** — a `loopback0` with an IPv4 → `(ip, "loopback0")`
- **Priority 2** — no loopback0, exactly one virtual interface with an IPv4 → `(ip, <label-or-name>)`, info-logged
- **Priority 3** — no loopback0, multiple virtual interfaces → the one tagged `out-of-band` with an IPv4 → `(ip, <label-or-name>)`, info-logged
- multiple virtual interfaces, none tagged `out-of-band` → warning logged (line 166), `(None, None)`
- loopback0 present but with **no** IPv4 → does not early-return, falls through to priority 2/3
- nothing resolvable → `(None, None)`

#### `_get_ipv4_from_interface(interface, netbox_client)` — `files/netbox/dnsmasq/metalbox_mode.py:173`

- a single IPv4 → returned without the prefix
- only IPv6 / no addresses → `None` (IPv6 filtered via the `"." in ip.address` test)
- multiple IPv4 → prefer the one inside an OOB-role prefix (via `get_all_oob_prefixes()`); none match → fall back to the first IPv4

#### `_get_physical_uplink_interfaces(device, netbox_client)` — `files/netbox/dnsmasq/metalbox_mode.py:207`

- interface carrying **both** `managed-by-osism` and `out-of-band` tags + a label + connected endpoints + enabled + not `mgmt_only` → its label appended
- missing either required tag → skipped (the `{...} <= tag_slugs` subset test)
- missing label / no connected endpoints / disabled / `mgmt_only` → skipped (one case each)
- `api.dcim.interfaces.filter` raising → warning logged (line 254), returns what was collected

#### `_get_dhcp_options_routed(metalbox_ip, prefix_mapping)` — `files/netbox/dnsmasq/metalbox_mode.py:259`

- per prefix → exactly three options `tag:{tag},3,{gateway}`, `tag:{tag},6,{metalbox_ip}`, `tag:{tag},42,{metalbox_ip}`, where `gateway == network_address + 1`

#### `_get_dynamic_hosts_routed(metalbox_ip, interface_name)` — `files/netbox/dnsmasq/metalbox_mode.py:280`

- returns exactly `["metalbox,{ip},{iface}", "metalbox.osism.xyz,{ip},{iface}"]`

---

#### `get_dynamic_hosts_for_metalbox(device, netbox_client)` — `files/netbox/dnsmasq/metalbox_mode.py:295`

- `get_oob_networks()` empty → `[]`
- `api.dcim.interfaces.filter` raising → warning logged (line 321), `[]`
- builds a VLAN-id → interface map from virtual interfaces carrying an `untagged_vlan` and IPs; for each OOB network with a matching `vlan.vid` and an IP inside the network's prefix → emits `"metalbox,{ip},{iface}"` **and** `"metalbox.osism.xyz,{ip},{iface}"`, then breaks on the first matching IP (line 386)
- a network without a `vlan`, or a VLAN with no matching interface → contributes nothing

#### `get_dhcp_options_for_metalbox(device, netbox_client)` — `files/netbox/dnsmasq/metalbox_mode.py:392`

- `api.dcim.interfaces.filter` raising → warning logged, `[]`
- `ipam.prefixes.filter` raising → `managed_vlan_ids = None` → the managed-VLAN filter is disabled (fallback, line 435)
- virtual + `managed-by-osism` + `untagged_vlan` whose vid **is** in `managed_vlan_ids` → emits option 6 (`tag:vlanXXX,6,ip`) and option 42 (`tag:vlanXXX,42,ip`) for the first IP
- a VLAN whose vid is **not** in `managed_vlan_ids` (when the set is populated) → interface skipped (line 457)
- VLAN in a group whose name contains `"routed"` → **additionally** emits option 3 (`tag:vlanXXX,3,{gateway}`) with `gateway == network_address + 1`; a non-routed VLAN emits only 6 & 42
- only the first IP per interface is used (the `break` at line 520); a gateway-calc exception → warning logged (line 515)

#### `process_devices(netbox_client, devices, all_devices)` — `files/netbox/dnsmasq/metalbox_mode.py:526`

> The densest function in the package; cover it with a few **focused** scenarios rather than one mega-test.

- no metalbox-role device in `devices` → early return (line 557), nothing written
- **mode detection**: the metalbox has VLAN interfaces → *bridged*; it has none → *routed* (line 564)
- **bridged scenario**: collects host / MAC / interface entries across `all_devices`, dedups by hostname / MAC / interface name, and writes the five `dnsmasq_*__metalbox` keys via `write_dnsmasq_to_device`; `dynamic_hosts` / `dhcp_options` come from `get_dynamic_hosts_for_metalbox` / `get_dhcp_options_for_metalbox`
- **routed scenario**: builds `prefix_mapping` from only the prefixes that actually have devices (sequential suffixes), derives each host's `set_tag` via `_get_set_tag_for_ip`, and uses the routed helpers plus `_get_physical_uplink_interfaces` for the bound interface list
- **switch tracking**: devices whose role slug is in `config.dnsmasq_switch_roles` are tracked separately and cached back to the metalbox device's custom field with the metalbox's own params **excluded** (line 798)
- a device with a MAC but no IP → contributes a MAC entry only; a non-MAC metalbox device in bridged mode still has its virtual interfaces cached (line 730)
- `update_device_custom_field` is called per processed device; a falsy return → warning logged

## Mocking hints

- **New conftest factories** (keyword-only, defaulted — tier 1-5 tests must keep passing):
  ```python
  def make_vlan_group(name):
      return SimpleNamespace(name=name)

  def make_vlan(vid, *, group=None):              # extends the tier-4 make_vlan(vid)
      return SimpleNamespace(vid=vid, group=group)

  def make_prefix(prefix, *, vlan=None):          # OOB network / prefix object
      return SimpleNamespace(prefix=prefix, vlan=vlan)
  ```
  `make_device` gains a keyword-only `device_type=None` (pass `SimpleNamespace(slug="server")` for the MAC-entry fallback). `make_fake_api` gains a `prefixes=()` argument backing `ipam.prefixes.filter(tag=...)` for `get_dhcp_options_for_metalbox`.
- **Config**: a small helper instead of importing the real `Config`:
  ```python
  def make_dnsmasq_config(tmp_path, *, reconciler_mode="manager",
                          dnsmasq_lease_time="28d", dnsmasq_switch_roles=("leaf",)):
      return SimpleNamespace(
          reconciler_mode=reconciler_mode, inventory_path=tmp_path,
          dnsmasq_lease_time=dnsmasq_lease_time,
          dnsmasq_switch_roles=list(dnsmasq_switch_roles),
      )
  ```
  `inventory_path` is the pytest `tmp_path` — the write paths (`group_vars/manager/...`, `host_vars/{hostname}/...`) are created and asserted against the real temp tree.
- **Fake `NetBoxClient`**: a `MagicMock()` whose `.api` is a `make_fake_api(...)` (or a nested `MagicMock` when `call_args` / `side_effect` matter) and whose façade methods are stubbed per scenario:
  ```python
  client = MagicMock()
  client.api = make_fake_api(interfaces=[...], ips_by_interface={...}, prefixes=[...])
  client.get_oob_networks.return_value = [make_prefix("192.0.2.0/24", vlan=make_vlan(100))]
  client.get_all_oob_prefixes.return_value = [...]
  client.get_device_oob_interface.return_value = ("192.0.2.10", "aa:bb:cc:dd:ee:ff", 100)
  client.update_device_custom_field.return_value = True
  ```
  `get_device_oob_interface` and `get_oob_networks` / `get_all_oob_prefixes` are tier-8 client methods — stub their return values here rather than re-deriving them.
- Use the **real** `get_inventory_hostname` (tier 1, #531) — the hostname-resolution and `host_vars` glob behaviour is exactly what `base.py` / `manager_mode.py` tests exercise.
- The **manager.py** dispatch test patches `manager.manager_handler` / `manager.metalbox_handler` on the instance with `MagicMock`s and asserts exactly one path is taken per mode; it needs no filesystem.
- `metalbox_mode.process_devices` (Part 7) is the densest path — build small interface / IP / OOB-network stubs and assert on the resulting `dnsmasq_data` dict and the custom-field writes; prefer one bridged, one routed and one switch-tracking scenario over a single combined test.
- Patch the per-module loggers (`dnsmasq.dhcp_config.logger`, `dnsmasq.base.logger`, `dnsmasq.interface_handler.logger`, `dnsmasq.manager_mode.logger`, `dnsmasq.metalbox_mode.logger`) only where a test asserts a specific warning / info message; otherwise let pytest capture them.

## Definition of Done

- [ ] `files/netbox/dnsmasq/dhcp_config.py` covered (`generate_dhcp_host_entry`, `generate_dhcp_mac_entry`, `write_dhcp_ranges`) in `test_dnsmasq_dhcp_config.py`
- [ ] `files/netbox/dnsmasq/base.py` covered (`write_dnsmasq_to_device`: existing-dir / existing-file-append / new-dir / multiple-match) in `test_dnsmasq_base.py`
- [ ] `files/netbox/dnsmasq/interface_handler.py` covered (`get_virtual_interfaces_for_dnsmasq`) in `test_dnsmasq_interface_handler.py`
- [ ] `files/netbox/dnsmasq/manager_mode.py` covered (`process_devices`, `process_devices_readonly`) in `test_dnsmasq_manager_mode.py`
- [ ] `files/netbox/dnsmasq/metalbox_mode.py` **Part 1** covered (helpers: `_build_prefix_tag_mapping`, `_index_to_suffix`, `_get_set_tag_for_ip`, `_get_metalbox_oob_virtual_interface`, `_get_ipv4_from_interface`, `_get_physical_uplink_interfaces`, `_get_dhcp_options_routed`, `_get_dynamic_hosts_routed`) in `test_dnsmasq_metalbox_mode.py`
- [ ] `files/netbox/dnsmasq/metalbox_mode.py` **Part 2** covered (`get_dynamic_hosts_for_metalbox`, `get_dhcp_options_for_metalbox`, `process_devices` bridged + routed + switch-tracking)
- [ ] `files/netbox/dnsmasq/manager.py` covered (`write_dnsmasq_config` mode dispatch, `write_dnsmasq_dhcp_ranges` pass-through) in `test_dnsmasq_manager.py`
- [ ] `tests/unit/netbox/conftest.py` extended backward-compatibly (`make_prefix`, `make_vlan(group=...)`, `make_vlan_group`, `make_device(device_type=...)`, `make_fake_api(prefixes=...)`, `make_dnsmasq_config`) — tier 1-5 tests still pass unchanged
- [ ] All listed branches covered (verify via `pytest --cov=files.netbox.dnsmasq`)
- [ ] `pytest tests/unit/netbox` passes locally; the tier 6 suite runs fast (filesystem writes go to `tmp_path`, no network, no real waiting)
- [ ] Tests do not leak state (each writes under its own `tmp_path`; `monkeypatch` undoes all patches)
- [ ] `flake8` and `black` remain green
- [ ] Zuul `container-image-inventory-reconciler-unit-tests` job passes

## Dependencies

- Part of #524 (meta)
- Builds on #525 (tier 0 foundation), #531 (tier 1 — `get_inventory_hostname`), #538 (tier 2 — `make_device` / `make_tag`), #543 (tier 3 — `make_ip` / `make_interface` / `make_fake_api`), #546 (tier 4 — `make_vlan`, `get_device_oob_interface` shape) and #550 (tier 5 — the extended `make_interface` with `type` / `label` / `connected_endpoints` / `enabled` / `mgmt_only`, reused for the dnsmasq interface gating)

### PR DESCRIPTION
**Pull request by berendt**
Closes #551

Adds tier 6 (dnsmasq) unit tests for the `files/netbox/dnsmasq/` package
and the conftest factories they need. **Test-only change — no
`files/netbox/**` source is touched.** Tests drive a `MagicMock`-backed
fake `NetBoxClient` (whose `.api` is `make_fake_api(...)`) and write into
pytest's `tmp_path` as the inventory root, asserting on the real temp
tree and on collaborator `call_args`.

## Change set, in commit order

1. **Add dnsmasq test factories to netbox conftest** — adds
   `make_iface_type` / `make_vlan` / `make_vlan_group` / `make_prefix` /
   `make_dnsmasq_config`, extends `make_device` (`device_type=`),
   `make_interface` (`type`/`label`/`name`/`untagged_vlan`/
   `connected_endpoints`/`enabled`, `mgmt_only` now defaulted) and
   `make_fake_api` (`prefixes=` backing `ipam.prefixes.filter(tag=...)`).
   Every addition is keyword-only and defaulted; tier 1-3 tests pass
   unchanged.
2. **dhcp_config** — host-entry formatting, MAC-entry priority chain,
   `write_dhcp_ranges` filesystem output under `tmp_path`.
3. **base** — `write_dnsmasq_to_device` across the four host_vars glob
   shapes (existing-dir / existing-file-append / new-dir / multi-match).
4. **interface_handler** — the static `get_virtual_interfaces_for_dnsmasq`
   gating.
5. **manager_mode** — `process_devices` / `process_devices_readonly`.
6. **manager** — `write_dnsmasq_config` mode dispatch and
   `write_dnsmasq_dhcp_ranges` pass-through.
7. **metalbox_mode Part 1** — pure / single-interface helpers.
8. **metalbox_mode Part 2** — the collectors plus focused bridged /
   routed / switch-tracking `process_devices` scenarios.

## Verification

- `pytest tests/unit` — 377 passed (259 pre-existing netbox tests
  unchanged + 118 new tier-6 tests).
- `pytest tests/unit/netbox --cov=dnsmasq` — 98% package coverage; all
  six modules covered, every branch listed in the issue exercised. The
  few remaining uncovered lines are defensive `except`/guard blocks not
  enumerated in the issue (covering them would require contrived mocks
  that do not reflect real NetBox behaviour).
- `flake8` and `black --check` green on every touched file.

## Deviation from the issue

The issue declares dependencies on #546 (tier 4) and #550 (tier 5), but
neither is merged to `main` yet — they exist only as unmerged branches.
Tier 6's interface/VLAN tests need their factories (`make_vlan`,
`make_iface_type`, the extended `make_interface`), so this PR ports the
minimal necessary factory shapes into the conftest itself (the fallback
the planning session anticipated) rather than blocking on the dependency
merge. All additions are backward compatible. If #546/#550 land first,
the overlapping factory definitions should be reconciled at merge time.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/unit/netbox/test_dnsmasq_metalbox_mode.py [hunk 1]
```diff
@@ -0,0 +1,1048 @@
...
+        assert written["dnsmasq_dhcp_hosts__metalbox"] == [
+            "aa:aa:aa:aa:aa:aa,metalbox,192.0.2.1",
+            "bb:bb:bb:bb:bb:bb,dup,192.0.2.10",
+        ]
+        assert written["dnsmasq_dhcp_macs__metalbox"] == [
+            "set:metal,aa:aa:aa:aa:aa:aa",
+            "set:server,bb:bb:bb:bb:bb:bb",
+        ]
+
+    def test_routed_uses_routed_helpers_and_set_tags(self, tmp_path, monkeypatch):
```
Comment (ideaship) [inline tests/unit/netbox/test_dnsmasq_metalbox_mode.py:826]: **Coverage gap that points at a real production gap (highest priority).** This stubs `_get_metalbox_oob_virtual_interface` to a valid `("192.0.2.254", "oob0")`, so the helper's real `(None, None)` return — which it explicitly produces when no usable metalbox virtual OOB IP/interface exists — is never exercised at the process level.

In `files/netbox/dnsmasq/metalbox_mode.py` the routed branch (around lines 760–773) feeds `metalbox_ip` / `metalbox_iface_name` straight into `_get_dynamic_hosts_routed`, `_get_dhcp_options_routed`, and `[metalbox_iface_name] + uplink_interfaces` with no None-guard anywhere on the path. With `(None, None)` that emits literal `None` into the written host_vars: `metalbox,None,None`, `tag:...,6,None` / `,42,None`, and `dnsmasq_interfaces__metalbox = [None, ...]` — a silently broken dnsmasq config.

Please add a routed `process_devices` test that drives the real helper (or fake API) through the `(None, None)` case, and add a guard that skips/aborts the routed write when the IP or interface is unresolved — unless emitting `None` is genuinely intended, which the surrounding code suggests it is not.

File: tests/unit/netbox/test_dnsmasq_metalbox_mode.py [hunk 2]
```diff
@@ -0,0 +1,1048 @@
...
+            {
+                1: ("192.0.2.1", "aa:aa:aa:aa:aa:aa", None),
+                2: ("192.0.2.10", "bb:bb:bb:bb:bb:bb", None),
+            }
+        )
+        client.update_device_custom_field.return_value = True
+
+        handler.process_devices(client, [metalbox], [metalbox, node])
+
+        write.assert_called_once()
```
Comment (ideaship) [inline tests/unit/netbox/test_dnsmasq_metalbox_mode.py:722]: **Question — please confirm intended behavior, then assert it.** This no-switch bridged path checks only `write_dnsmasq_to_device`; it never inspects `client.update_device_custom_field`. In production the per-device loop caches the metalbox's *own* params (around `metalbox_mode.py:721`), and the switch-only overwrite (around lines 803–818) is guarded by `if switch_dhcp_hosts or switch_dhcp_macs`. With no switch devices that overwrite is skipped, so the metalbox custom field retains its own generated params — seemingly contradicting the "discard metalbox own parameters" comment near line 798.

The "metalbox own excluded" behavior is only asserted in the switch-present test; this reachable no-switch state is uncovered. (`test_non_mac_metalbox_caches_virtual_interfaces` asserts the caching call only for the no-MAC branch.) Please confirm whether retaining own-params here is intended, and add assertions on the final `update_device_custom_field` call so the state is locked in either way.

File: tests/unit/netbox/test_dnsmasq_metalbox_mode.py [hunk 3]
```diff
@@ -0,0 +1,1048 @@
...
+            [iface],
+            {1: [make_ip("192.0.2.999/24")]},  # malformed -> gateway calc raises
+            prefixes=[make_prefix("192.0.2.0/24", vlan=make_vlan(100))],
+        )
+
+        result = handler.get_dhcp_options_for_metalbox(
+            make_device(2, "metalbox"), client
+        )
+
+        # 6 and 42 still emitted from the raw string; the gateway option is not.
```
Comment (ideaship) [inline tests/unit/netbox/test_dnsmasq_metalbox_mode.py]: Minor / non-blocking: worth a one-line note that this is a *characterization* of existing behavior, not intended validation. In `get_dhcp_options_for_metalbox` only option 3 (gateway) is guarded by the `ip_network()` try/except; options 6 and 42 come from an unvalidated `split("/")[0]`, so a malformed `192.0.2.999` leaks straight through. Fine to keep as a regression pin — just flag it so a future reader doesn't mistake this for "invalid IPs are validated."

File: tests/unit/netbox/test_dnsmasq_interface_handler.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    def test_qualifying_interface_falls_back_to_name(self):
+        assert _call([_virtual(label=None, name="eth0")]) == ["eth0"]
+
+    def test_multiple_qualifying_interfaces_preserve_order(self):
+        iface1 = _virtual(id=1, label="oob1")
+        iface2 = _virtual(id=2, label="oob2")
+        assert _call([iface1, iface2]) == ["oob1", "oob2"]
+
+    def test_callable_as_static_method_without_instance(self):
```
Comment (ideaship) [inline tests/unit/netbox/test_dnsmasq_interface_handler.py]: Minor / non-blocking: this duplicates `test_qualifying_interface_prefers_label_and_debug_logs_vid` — same `_call([_virtual()])` → `["oob1"]`. Every test in this class already invokes `_call` on the class, so "without an instance" isn't a distinct path. Suggest dropping or merging.

File: tests/unit/netbox/conftest.py [hunk 5]
```diff
@@ -94,5 +164,30 @@ def make_fake_api(interfaces=(), ips_by_interface=None):
                     ips_by_interface.get(interface_id, [])
                 ),
             ),
+            prefixes=SimpleNamespace(
+                filter=lambda tag: list(prefixes),
```
Comment (ideaship) [inline tests/unit/netbox/conftest.py]: Minor / non-blocking: this faked `prefixes.filter` ignores its `tag` argument, so the `tag=["managed-by-osism"]` filter in `get_dhcp_options_for_metalbox` is never exercised — a regression dropping that filter would pass silently. Acceptable as a fake limitation, but consider recording the passed `tag` and asserting it in at least one test so the filter is pinned.

**Review by ideaship — CHANGES_REQUESTED** [review]

**Review by ideaship — APPROVED** [review]
